File: app/util/java/ExternalCourseHandler.java
package util.java;

import com.google.inject.ImplementedBy;
import models.Course;
import models.User;
import play.mvc.Result;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletionStage;

@ImplementedBy(ExternalCourseHandlerImpl.class)
public interface ExternalCourseHandler {
    CompletionStage<List<Course>> getCoursesByCode(User user, String code) throws IOException;
    CompletionStage<Result> updateCourses() throws IOException;
    CompletionStage<Collection<String>> getPermittedCourses(User user) throws IOException;
}

File: app/services/exam/ExternalCourseHandler.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.exam

import com.google.inject.ImplementedBy
import models.exam.Course
import models.user.User

import scala.concurrent.Future

@ImplementedBy(classOf[ExternalCourseHandlerImpl])
trait ExternalCourseHandler:
  def getCoursesByCode(user: User, code: String): Future[Set[Course]]
  def getPermittedCourses(user: User): Future[Set[String]]
